server/index.js

```diff
@@ -113,33 +115,42 @@ async function handleAlert(location, alertType) {
   }
 }
 
-app.post('/api/sensorEvent', Validator.body(['coreid', 'event']).exists(), async (request, response) => {
+app.post('/api/sensorEvent', Validator.body(['coreid', 'event', 'api_key']).exists(), async (request, response) => {
   try {
     const validationErrors = Validator.validationResult(request).formatWith(helpers.formatExpressValidationErrors)
 
     if (validationErrors.isEmpty()) {
-      let alertType
-      const coreId = request.body.coreid
-      const sensorEvent = request.body.event
-      if (sensorEvent === SENSOR_EVENT.DURATION) {
-        alertType = ALERT_TYPE.SENSOR_DURATION
-      } else if (sensorEvent === SENSOR_EVENT.STILLNESS) {
-        alertType = ALERT_TYPE.SENSOR_STILLNESS
+      const apiKey = request.body.api_key
+
+      if (webhookAPIKey === apiKey) {
+        let alertType
+        const coreId = request.body.coreid
+        const sensorEvent = request.body.event
+        if (sensorEvent === SENSOR_EVENT.DURATION) {
+          alertType = ALERT_TYPE.SENSOR_DURATION
+        } else if (sensorEvent === SENSOR_EVENT.STILLNESS) {
+          alertType = ALERT_TYPE.SENSOR_STILLNESS
+        } else {
+          const errorMessage = `Bad request to ${request.path}: Invalid event type`
+          helpers.logError(errorMessage)
+        }
+        const location = await db.getLocationFromParticleCoreID(coreId)
+        if (!location) {
+          const errorMessage = `Bad request to ${request.path}: no location matches the coreID ${coreId}`
+          helpers.logError(errorMessage)
+          // Must send 200 so as not to be throttled by Particle (ref: https://docs.particle.io/reference/device-cloud/webhooks/#limits)
+          response.status(200).json(errorMessage)
+        } else {
+          if (location.client.isActive && location.isActive) {
+            await handleAlert(location, alertType)
+          }
+          response.status(200).json('OK')
+        }
       } else {
-        const errorMessage = `Bad request to ${request.path}: Invalid event type`
-        helpers.logError(errorMessage)
-      }
-      const location = await db.getLocationFromParticleCoreID(coreId)
-      if (!location) {
-        const errorMessage = `Bad request to ${request.path}: no location matches the coreID ${coreId}`
+        const errorMessage = `Access not allowed: incorrect key ${apiKey}`
```

Comment:
For this real error log, I'd actually leave out the incorrect key value that was sent. 

In general, we want to log enough data that we can debug problems, but not so much that any hacker would be able to learn more about our system. So it's possible that someone provides an API key that is close, like they just did a typo. In which case, that typo would now be logged and a potential hacker might guess what the problem was... maybe extra whitespace or capitalization and then be able to know how API key.

Reply:
ohhhh that's a great point, thank you! i'll update the error log accordingly.